Flask DocumentBlueprint: Add to spec all views defined in the blueprint.

JavierLuna opened the pull request:
# Flask DocumentBlueprint
Is a blueprint  which adds all views defined in it to the spec. You can ommit one view from being added to the spec setting `documented=False` in the Blueprint's `route` function (see usage below).

## Usage
```python

from flask import Flask
from flask.views import MethodView

app = Flask(__name__)

documented_blueprint = DocumentedBlueprint('gistapi', __name__)

@documented_blueprint.route('/gists/<gist_id>')
def gist_detail(gist_id):
    '''Gist detail view.
    ---
    x-extension: metadata
    get:
        responses:
            200:
                schema:
                    $ref: '#/definitions/Gist'
    '''
    return 'detail for gist {}'.format(gist_id)


@documented_blueprint.route('/repos/<repo_id>', documented=False)
def repo_detail(repo_id):
    '''This endpoint won't be documented
    ---
    x-extension: metadata
    get:
        responses:
            200:
                schema:
                    $ref: '#/definitions/Repo'
    '''
    return 'detail for repo {}'.format(repo_id)

app.register_blueprint(documented_blueprint)

print(spec.to_dict()['paths'])
# {'/gists/{gist_id}': {'get': {'responses': {200: {'schema': {'$ref': '#/definitions/Gist'}}}},
#                  'x-extension': 'metadata'}}

```

Has a test coverage of 100%.

sloria commented on .gitignore:
```diff
@@ -100,5 +100,9 @@ venv.bak/
 /site
 
+# Editors
+.idea/
```
Drive-by: This belongs in your [global gitignore file](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore).

JavierLuna replied:
Done!

sloria commented:
@tinproject @ergo Would you be up for reviewing this API, since you had some thoughts on #11?

tinproject commented:
I hope I get more time on the weekend to take a proper look, this is just a quick read.

I'm not convinced of the extension by subclassing approach. I'm not sure what happen with MethodViews or other extensions of flask. Also if I remember well one problem is that one view function could have more than one route and should documented for every route, but I'm not very familiar with new openapi spec.

I'll try to try this PR on the weekend and write some code with it to clarify my thoughts.

pcaro commented:
I like the idea implementation @JavierLuna but I found a little glitches:
- It's not in the example or module documentation, but you need to pass the `spec` object to the `DocumentedBlueprint` constructor. 
- The route decorator appends a nested unnecessary function call (nested decorator)

JavierLuna commented:
I'll fix them as soon as I have a bit of free time.
Thanks @pcaro !

tinproject commented:
I finally get some hours and take a look at this (it's been a year since I made something with apispec so I have to refresh my mind)

I've created a dumb project (at https://github.com/tinproject/apispec-webframeworks-test) and try the PR with different approach to add routes.

First had to said that the code of Javier is good, it only lacks support for Blueprint [`add_url_rule`](http://flask.pocoo.org/docs/1.0/api/#flask.Blueprint.add_url_rule) to be complete.

I saw two drawbacks:
1.- The subclassing approach force to inject the spec object, it could be a problem when having multiple separated blueprints, you will need some tricks to pass the spec object around and avoid circular references. Also add some toil to tests.
2.- Although it fulfills the issue #11 scope, it's more of a patch for this concrete case. It will be good to fix the [FlaskPlugin](https://github.com/marshmallow-code/apispec-webframeworks/blob/master/apispec_webframeworks/flask.py#L93) one and for all, taking into account Flask/Werkzeug url routing, and also fix #14 with a complete solution. The good part of this PR is that is contained and could be easily deprecated and removed in the future if a proper solution is added finally to the project.

On the project referenced above I played around url routing in flask, trying to show the map/rules/endpoints/view_functions relationships if you want to take a peek.

JavierLuna commented on apispec_webframeworks/flask.py:
```diff
@@ -114,3 +153,38 @@ def path_helper(self, operations, view, **kwargs):
+        self.documented_view_functions = defaultdict(list)
+        self.spec = spec
+
+    def route(self, rule, documented=True, **options):
```
I overrode this function to include the `documented=True` kwarg, as I like declaring them explicitly.

Do you like it this way or shall I remove this?

JavierLuna commented:
Thanks for checking it out and for the suggestions @tinproject !

I refactored a little the implementation and added some tests reflecting your project's cases.
`add_url_rule` is now supported :+1: 

Regarding your drawbacks:

1.- How would you implement it? The spec object is only needed when registering the application, maybe have it as `spec=None` in the `__init__` method and then have some kind of `attach(spec)`?

2.- I looked around and it seems that it cannot be changed in [this project](https://github.com/marshmallow-code/apispec-webframeworks/blob/01f29ff84456a72a66fbf8c5526a0f4c81565898/apispec_webframeworks/flask.py#L103) and [a change in the Apispec project](https://github.com/marshmallow-code/apispec/blob/5e4b651d06b964bad8fe60393c4927310e472e40/apispec/core.py#L292) will be needed. `path_helper` is expected by the `APISpec` object to return a single path, but in this case it is necessary to send multiple ones.
Shall I start working on it? What are your thoughts on the issue?

pcaro commented on apispec_webframeworks/flask.py:
```diff
@@ -114,3 +153,38 @@ def path_helper(self, operations, view, **kwargs):
+        if documented:
+            self.documented_view_functions[rule].append(view_func)
+
+    def register(self, app, options, first_registration=False):
```
For my usercase I prefer to pass the spec in the register (or add_url_rule method) instead of the constructor.

JavierLuna replied:
That could be a solution, yeah!

`app.register_blueprint(doc_blueprint, spec=spec)`

I'd still add the option of passing it through the `__init__` as a kwarg and add an `attach(spec)` method, to add variety of uses.

What do you think @tinproject ?

jshwelz commented:
any update on this?

caffeinatedMike commented:
@sloria @pcaro Can you re-review this PR? I'm in need of a solution that allows creating individual specs, one per blueprint, that are version-specific. This seems like the perfect solution.
